File: src/commands/sync-back.ts
import { existsSync, readFileSync, writeFileSync, mkdirSync, readdirSync, rmSync } from 'fs';
import { join, relative, dirname } from 'path';
import { tmpdir } from 'os';
import { Manifest, loadIgnorePatterns, isIgnored } from '../lib/manifest.js';
import { git, ghCli, getCurrentBranch, getRepoName, isGitRepo } from '../lib/git.js';
import { getAllhandsRoot, UPSTREAM_REPO } from '../lib/paths.js';

const PROTECTED_BRANCHES = new Set(['main', 'master', 'develop', 'staging', 'production']);

function walkDir(dir: string, callback: (filePath: string) => void): void {
  if (!existsSync(dir)) return;
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      walkDir(fullPath, callback);
    } else if (entry.isFile()) {
      callback(fullPath);
    }
  }
}

function getChangedManagedFiles(
  targetRoot: string,
  allhandsRoot: string,
  manifest: Manifest,
  ignorePatterns: string[]
): string[] {
  const changes: string[] = [];

  for (const relPath of manifest.getDistributableFiles()) {
    if (isIgnored(relPath, ignorePatterns)) continue;

    const sourceFile = join(allhandsRoot, relPath);
    const targetFile = join(targetRoot, relPath);

    if (!existsSync(targetFile)) continue;

    if (!existsSync(sourceFile)) {
      changes.push(relPath);
      continue;
    }

    const sourceContent = readFileSync(sourceFile);
    const targetContent = readFileSync(targetFile);
    if (!sourceContent.equals(targetContent)) {
      changes.push(relPath);
    }
  }

  return changes;
}

function getNewFilesInManagedDirs(
  targetRoot: string,
  allhandsRoot: string,
  manifest: Manifest,
  ignorePatterns: string[]
): string[] {
  const newFiles: string[] = [];

  const claudeDir = join(targetRoot, '.claude');
  if (existsSync(claudeDir)) {
    walkDir(claudeDir, (targetFile) => {
      const relPath = relative(targetRoot, targetFile);

      if (isIgnored(relPath, ignorePatterns)) return;

      const sourceFile = join(allhandsRoot, relPath);
      if (!existsSync(sourceFile) && manifest.isDistributable(relPath)) {
        newFiles.push(relPath);
      }
    });
  }

  return newFiles;
}

function cloneAllhandsToTemp(): string | null {
  const tempDir = join(tmpdir(), `allhands-sync-${Date.now()}`);

  console.log(`Cloning ${UPSTREAM_REPO} to temp directory...`);
  const result = ghCli(['repo', 'clone', UPSTREAM_REPO, tempDir, '--', '--depth=1'], process.cwd());

  if (!result.success) {
    console.error(`Failed to clone: ${result.stderr}`);
    return null;
  }

  return tempDir;
}

function cleanupTempDir(tempDir: string): void {
  try {
    rmSync(tempDir, { recursive: true, force: true });
  } catch (e) {
    // Ignore cleanup errors
  }
}

export interface SyncBackOptions {
  auto?: boolean;
  list?: boolean;
}

export async function cmdSyncBack(options: SyncBackOptions = {}): Promise<number> {
  const { auto = false, list = false } = options;
  const targetRoot = process.cwd();

  if (!isGitRepo(targetRoot)) {
    if (!list) console.error('Error: Not in a git repository');
    return 1;
  }

  const currentBranch = getCurrentBranch(targetRoot);

  if (!currentBranch) {
    if (auto) return 0;
    if (!list) {
      console.error('Error: Could not determine current branch');
      console.error('Ensure you have at least one commit and are on a branch');
    }
    return 1;
  }

  if (auto && !PROTECTED_BRANCHES.has(currentBranch)) {
    return 0;
  }

  // Get allhands source for comparison (from package or ALLHANDS_PATH)
  const sourceRoot = getAllhandsRoot();

  if (!existsSync(join(sourceRoot, '.allhands-manifest.json'))) {
    if (!list) console.error(`Error: Manifest not found at ${sourceRoot}`);
    return 1;
  }

  const manifest = new Manifest(sourceRoot);
  const ignorePatterns = loadIgnorePatterns(targetRoot);

  // Find changed files
  const changedFiles = getChangedManagedFiles(targetRoot, sourceRoot, manifest, ignorePatterns);
  const newFiles = getNewFilesInManagedDirs(targetRoot, sourceRoot, manifest, ignorePatterns);
  const filesToSync = [...changedFiles, ...newFiles];

  // --list mode: just print files that would sync, one per line
  if (list) {
    for (const f of filesToSync) {
      console.log(f);
    }
    return 0;
  }

  if (filesToSync.length === 0) {
    console.log('No changes to sync back');
    return 0;
  }

  // Check gh auth (only needed for actual sync, not --list)
  const authResult = ghCli(['auth', 'status'], targetRoot);
  if (!authResult.success) {
    console.error('Error: Not authenticated with GitHub CLI');
    console.error('Run: gh auth login');
    return 1;
  }

  const repoName = getRepoName(targetRoot);
  const prBranch = `sync/${repoName}/${currentBranch}`;

  console.log(`\n📋 Syncing ${filesToSync.length} file(s) to ${UPSTREAM_REPO}`);
  console.log(`   Branch: ${prBranch}\n`);

  for (const f of filesToSync) {
    const isNew = newFiles.includes(f);
    console.log(`   ${isNew ? '[NEW]' : '[MOD]'} ${f}`);
  }
  console.log();

  // Always clone to temp directory to avoid conflicts with local staged changes
  const allhandsClone = cloneAllhandsToTemp();
  if (!allhandsClone) {
    return 1;
  }

  try {
    // Fetch latest and checkout/create branch
    git(['fetch', 'origin'], allhandsClone);

    const branchCheck = git(['ls-remote', '--heads', 'origin', prBranch], allhandsClone);
    const branchExists = branchCheck.stdout.includes(prBranch);

    if (branchExists) {
      console.log(`Checking out existing branch: ${prBranch}`);
      git(['checkout', '-B', prBranch, `origin/${prBranch}`], allhandsClone);
    } else {
      console.log(`Creating new branch: ${prBranch}`);
      git(['checkout', '-B', prBranch, 'origin/main'], allhandsClone);
    }

    // Copy files from target to allhands clone
    for (const relPath of filesToSync) {
      const source = join(targetRoot, relPath);
      const dest = join(allhandsClone, relPath);

      if (existsSync(source)) {
        mkdirSync(dirname(dest), { recursive: true });
        writeFileSync(dest, readFileSync(source));
      }
    }

    // Stage and check for changes
    git(['add', '-A'], allhandsClone);

    const diffResult = git(['diff', '--cached', '--quiet'], allhandsClone);
    if (diffResult.success) {
      console.log('✅ No changes to commit (files already synced)');
      return 0;
    }

    // Commit
    const commitMsg = `sync: ${repoName}/${currentBranch}`;
    git(['commit', '-m', commitMsg], allhandsClone);

    // Push
    console.log(`Pushing to origin/${prBranch}...`);
    const pushResult = git(['push', '-u', 'origin', prBranch], allhandsClone);
    if (!pushResult.success) {
      console.error(`Push failed: ${pushResult.stderr}`);
      return 1;
    }

    // Create or check PR
    const prViewResult = ghCli(['pr', 'view', prBranch, '--repo', UPSTREAM_REPO], allhandsClone);

    if (prViewResult.success) {
      console.log(`\n✅ PR updated: ${prBranch}`);
      // Get PR URL
      const prUrlResult = ghCli(['pr', 'view', prBranch, '--repo', UPSTREAM_REPO, '--json', 'url', '-q', '.url'], allhandsClone);
      if (prUrlResult.success) {
        console.log(`   ${prUrlResult.stdout}`);
      }
    } else {
      // Create new PR
      const fileList = filesToSync.map(f => `- \`${f}\``).join('\n');
      const prBody = `Automated sync-back from target repository.

**Source repo:** ${repoName}
**Branch:** \`${currentBranch}\`
**Files changed:** ${filesToSync.length}

${fileList}`;

      const prCreateResult = ghCli([
        'pr', 'create',
        '--repo', UPSTREAM_REPO,
        '--title', `sync: ${repoName}/${currentBranch}`,
        '--body', prBody,
        '--base', 'main',
        '--head', prBranch,
      ], allhandsClone);

      if (prCreateResult.success) {
        console.log(`\n✅ PR created: ${prCreateResult.stdout}`);
      } else {
        console.error(`\n❌ PR creation failed: ${prCreateResult.stderr}`);
        return auto ? 0 : 1;
      }
    }

    return 0;
  } finally {
    console.log('Cleaning up temp directory...');
    cleanupTempDir(allhandsClone);
  }
}

File: scripts/allhands/sync_back.py
"""Sync changes back to allhands as PR."""

import os
import subprocess
import sys
from pathlib import Path
from typing import List

from .manifest import Manifest, is_ignored, load_ignore_patterns


# Protected branches trigger auto sync-back via post-merge hook
# All branches can use manual sync-back
PROTECTED_BRANCHES = {"main", "master", "develop", "staging", "production"}


def get_current_branch(repo_path: Path) -> str:
    """Get current git branch name."""
    result = subprocess.run(
        ["git", "rev-parse", "--abbrev-ref", "HEAD"],
        cwd=repo_path,
        capture_output=True,
        text=True,
    )
    return result.stdout.strip() if result.returncode == 0 else ""


def get_repo_name(repo_path: Path) -> str:
    """Get repository name from git remote or directory."""
    result = subprocess.run(
        ["git", "remote", "get-url", "origin"],
        cwd=repo_path,
        capture_output=True,
        text=True,
    )
    if result.returncode == 0:
        url = result.stdout.strip()
        # Extract repo name from URL
        name = url.rstrip("/").split("/")[-1]
        if name.endswith(".git"):
            name = name[:-4]
        return name
    return repo_path.name


def get_changed_managed_files(
    target_root: Path,
    allhands_root: Path,
    manifest: Manifest,
    ignore_patterns: List[str],
) -> List[Path]:
    """Get managed files that differ from source (sync-back candidates)."""
    changes = []

    for rel_path in manifest.get_distributable_files():
        str_path = str(rel_path)

        # Skip ignored files
        if is_ignored(str_path, ignore_patterns):
            continue

        source_file = allhands_root / rel_path
        target_file = target_root / rel_path

        if not target_file.exists():
            continue

        if not source_file.exists():
            # New file in target - sync-back candidate
            changes.append(rel_path)
            continue

        if source_file.read_bytes() != target_file.read_bytes():
            changes.append(rel_path)

    return changes


def get_new_files_in_managed_dirs(
    target_root: Path,
    allhands_root: Path,
    manifest: Manifest,
    ignore_patterns: List[str],
) -> List[Path]:
    """Find new files in target that don't exist in source but match distribute patterns."""
    new_files = []

    # Check .claude/ directory for new files
    claude_dir = target_root / ".claude"
    if claude_dir.exists():
        for target_file in claude_dir.rglob("*"):
            if not target_file.is_file():
                continue

            rel_path = target_file.relative_to(target_root)
            str_path = str(rel_path)

            # Skip ignored
            if is_ignored(str_path, ignore_patterns):
                continue

            # Check if it's a new file (not in source)
            source_file = allhands_root / rel_path
            if not source_file.exists() and manifest.is_distributable(str_path):
                new_files.append(rel_path)

    return new_files


def create_or_update_pr(
    allhands_root: Path,
    target_root: Path,
    files_to_sync: List[Path],
    branch_name: str,
) -> bool:
    """Create or update PR in allhands repo with synced changes."""
    repo_name = get_repo_name(target_root)
    pr_branch = f"{repo_name}/{branch_name}"

    # Check if branch exists
    result = subprocess.run(
        ["git", "rev-parse", "--verify", pr_branch],
        cwd=allhands_root,
        capture_output=True,
    )
    branch_exists = result.returncode == 0

    if not branch_exists:
        # Create new branch from main
        subprocess.run(
            ["git", "checkout", "-b", pr_branch, "main"],
            cwd=allhands_root,
            capture_output=True,
        )
    else:
        subprocess.run(
            ["git", "checkout", pr_branch],
            cwd=allhands_root,
            capture_output=True,
        )

    # Copy files from target to allhands
    for rel_path in files_to_sync:
        source = target_root / rel_path
        dest = allhands_root / rel_path

        if source.exists():
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_bytes(source.read_bytes())
        elif dest.exists():
            dest.unlink()

    # Stage and commit
    subprocess.run(
        ["git", "add", "-A"],
        cwd=allhands_root,
        capture_output=True,
    )

    result = subprocess.run(
        ["git", "diff", "--cached", "--quiet"],
        cwd=allhands_root,
    )
    if result.returncode == 0:
        subprocess.run(["git", "checkout", "main"], cwd=allhands_root, capture_output=True)
        return True

    commit_msg = f"sync: {repo_name}/{branch_name}"
    subprocess.run(
        ["git", "commit", "-m", commit_msg],
        cwd=allhands_root,
        capture_output=True,
    )

    # Push branch
    result = subprocess.run(
        ["git", "push", "-u", "origin", pr_branch],
        cwd=allhands_root,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(f"Warning: Push failed: {result.stderr}", file=sys.stderr)

    # Create PR via gh CLI
    result = subprocess.run(
        ["gh", "pr", "view", pr_branch],
        cwd=allhands_root,
        capture_output=True,
    )
    pr_exists = result.returncode == 0

    if not pr_exists:
        result = subprocess.run(
            [
                "gh", "pr", "create",
                "--title", f"Sync from {repo_name}/{branch_name}",
                "--body", f"Automated sync-back from target repository.\n\nBranch: `{branch_name}`\nFiles: {len(files_to_sync)}",
                "--base", "main",
            ],
            cwd=allhands_root,
            capture_output=True,
            text=True,
        )
        if result.returncode == 0:
            print(f"PR: {result.stdout.strip()}")
        else:
            print(f"PR creation failed: {result.stderr}", file=sys.stderr)
    else:
        print(f"PR updated: {pr_branch}")

    # Return to main
    subprocess.run(["git", "checkout", "main"], cwd=allhands_root, capture_output=True)

    return True


def cmd_sync_back(auto: bool = False) -> int:
    """Sync changes back to allhands as PR.

    Any branch can sync back - each creates its own PR: [repo]/[branch]
    Auto mode (hook) triggers on protected branch merges.
    Manual mode works on any branch.

    Args:
        auto: Non-interactive mode (for hooks) - silent on errors
    """
    target_root = Path.cwd().resolve()

    # Check we're in a git repo
    if not (target_root / ".git").exists():
        print("Error: Not in a git repository", file=sys.stderr)
        return 1

    current_branch = get_current_branch(target_root)

    # Validate branch name
    if not current_branch:
        if auto:
            return 0  # Silent exit - no branch (detached HEAD or new repo)
        print("Error: Could not determine current branch", file=sys.stderr)
        print("Ensure you have at least one commit and are on a branch", file=sys.stderr)
        return 1

    # In auto mode (hook), only run on protected branches
    # Manual mode (no --auto flag) works on ANY branch
    if auto and current_branch not in PROTECTED_BRANCHES:
        # Silent exit - hook only triggers on protected branches
        return 0

    allhands_root_env = os.environ.get("ALLHANDS_PATH")
    if not allhands_root_env:
        if auto:
            print("ALLHANDS_PATH not set - skipping sync-back", file=sys.stderr)
            return 0
        print("Error: ALLHANDS_PATH environment variable not set", file=sys.stderr)
        return 1

    allhands_root = Path(allhands_root_env).resolve()

    if not (allhands_root / ".allhands-manifest.json").exists():
        print(f"Error: Manifest not found at {allhands_root}", file=sys.stderr)
        return 1

    manifest = Manifest(allhands_root)
    ignore_patterns = load_ignore_patterns(target_root)

    # Find changed files (all non-ignored differences sync back)
    changed_files = get_changed_managed_files(
        target_root, allhands_root, manifest, ignore_patterns
    )

    # Find new files in managed dirs
    new_files = get_new_files_in_managed_dirs(
        target_root, allhands_root, manifest, ignore_patterns
    )

    files_to_sync = changed_files + new_files

    if not files_to_sync:
        print("No changes to sync back")
        return 0

    repo_name = get_repo_name(target_root)
    pr_branch = f"{repo_name}/{current_branch}"

    print(f"\nSyncing {len(files_to_sync)} file(s) back to claude-all-hands...")
    print(f"PR branch: {pr_branch}")
    print(f"Files:")
    for f in files_to_sync:
        print(f"  → {f}")
    print(f"\nTo exclude files, add patterns to .allhandsignore")

    # Create PR (no confirmation needed)
    try:
        success = create_or_update_pr(
            allhands_root, target_root, files_to_sync, current_branch
        )
        return 0 if success else 1
    except Exception as e:
        print(f"Error: Sync-back failed: {e}", file=sys.stderr)
        return 1 if not auto else 0  # Don't fail hooks
